File: services/iceberg-catalog-service/src/handlers/auth/api_tokens.rs
//! `POST /v1/iceberg-clients/api-tokens` — mint long-lived bearer
//! tokens "tied to your user" (per the auth doc).
//!
//! The response surfaces the secret exactly once. Callers must
//! authenticate as a Foundry user (regular JWT). The minted token is
//! `ofty_<64 hex>` and is recognized by the bearer extractor.

use axum::Json;
use axum::extract::State;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::AppState;
use crate::audit;
use crate::domain::token as api_token_domain;
use crate::handlers::auth::bearer::AuthenticatedPrincipal;
use crate::handlers::errors::ApiError;

#[derive(Debug, Deserialize)]
pub struct CreateApiTokenRequest {
    pub name: String,
    #[serde(default)]
    pub scopes: Option<Vec<String>>,
    /// Optional TTL in seconds. When omitted defaults to the service's
    /// `long_lived_token_ttl_secs` (90 days).
    #[serde(default)]
    pub ttl_secs: Option<i64>,
}

#[derive(Debug, Serialize)]
pub struct CreateApiTokenResponse {
    pub id: Uuid,
    pub name: String,
    pub token_hint: String,
    pub scopes: Vec<String>,
    pub expires_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub raw_token: String,
}

pub async fn create_api_token(
    State(state): State<AppState>,
    principal: AuthenticatedPrincipal,
    Json(body): Json<CreateApiTokenRequest>,
) -> Result<Json<CreateApiTokenResponse>, ApiError> {
    let user_id = Uuid::parse_str(&principal.subject)
        .map_err(|_| ApiError::BadRequest("subject is not a UUID".to_string()))?;

    let scopes = body
        .scopes
        .unwrap_or_else(|| {
            vec![
                "api:iceberg-read".to_string(),
                "api:iceberg-write".to_string(),
            ]
        });

    let ttl = body.ttl_secs.or(Some(state.iceberg.long_lived_token_ttl_secs));
    let issued = api_token_domain::issue(
        &state.iceberg.db,
        user_id,
        &body.name,
        scopes.clone(),
        ttl,
    )
    .await
    .map_err(|err| ApiError::Internal(err.to_string()))?;

    audit::api_token_created(user_id, issued.record.id, &scopes);

    Ok(Json(CreateApiTokenResponse {
        id: issued.record.id,
        name: issued.record.name,
        token_hint: issued.record.token_hint,
        scopes: issued.record.scopes,
        expires_at: issued.record.expires_at,
        created_at: issued.record.created_at,
        raw_token: issued.raw_token,
    }))
}

File: services/iceberg-catalog-service/internal/handlers/auth/api_tokens.go
package auth

import (
	"context"
	"encoding/json"
	"net/http"
	"time"

	"github.com/google/uuid"

	authmw "github.com/openfoundry/openfoundry-go/libs/auth-middleware"
	"github.com/openfoundry/openfoundry-go/services/iceberg-catalog-service/internal/audit"
	tokendomain "github.com/openfoundry/openfoundry-go/services/iceberg-catalog-service/internal/domain/token"
)

// IssueAPITokenStore is the contract `CreateAPITokenHandler` needs from
// the data layer. Implementations live in `internal/repo`; tests
// substitute fakes.
type IssueAPITokenStore interface {
	IssueAPIToken(ctx context.Context, userID uuid.UUID, name string, scopes []string, expiresAt *time.Time) (*tokendomain.APIToken, string, error)
}

// CreateAPITokenRequest is the body of POST /v1/iceberg-clients/api-tokens.
type CreateAPITokenRequest struct {
	Name    string   `json:"name"`
	Scopes  []string `json:"scopes,omitempty"`
	TTLSecs *int64   `json:"ttl_secs,omitempty"`
}

// CreateAPITokenResponse is the response body. `RawToken` is shown to
// the caller exactly once — the catalog never recovers it.
type CreateAPITokenResponse struct {
	ID        uuid.UUID  `json:"id"`
	Name      string     `json:"name"`
	TokenHint string     `json:"token_hint"`
	Scopes    []string   `json:"scopes"`
	ExpiresAt *time.Time `json:"expires_at"`
	CreatedAt time.Time  `json:"created_at"`
	RawToken  string     `json:"raw_token"`
}

// CreateAPITokenHandler returns the chi-compatible HTTP handler for the
// POST endpoint. Authenticates as a Foundry user (regular Foundry JWT
// via the libs/auth-middleware chain) and mints an `ofty_*` token tied
// to that user.
func CreateAPITokenHandler(store IssueAPITokenStore, defaultTTLSecs int64) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		caller, ok := authmw.FromContext(r.Context())
		if !ok {
			writeJSONErr(w, http.StatusUnauthorized, "authentication required")
			return
		}
		var body CreateAPITokenRequest
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			writeJSONErr(w, http.StatusBadRequest, "invalid body")
			return
		}
		if body.Name == "" {
			writeJSONErr(w, http.StatusBadRequest, "name is required")
			return
		}
		scopes := body.Scopes
		if len(scopes) == 0 {
			scopes = []string{"api:iceberg-read", "api:iceberg-write"}
		}
		ttl := defaultTTLSecs
		if body.TTLSecs != nil {
			ttl = *body.TTLSecs
		}
		var expiresAt *time.Time
		if ttl > 0 {
			t := time.Now().UTC().Add(time.Duration(ttl) * time.Second)
			expiresAt = &t
		}
		record, raw, err := store.IssueAPIToken(r.Context(), caller.Sub, body.Name, scopes, expiresAt)
		if err != nil {
			writeJSONErr(w, http.StatusInternalServerError, err.Error())
			return
		}
		audit.APITokenCreated(caller.Sub, record.ID, scopes)
		writeJSON(w, http.StatusOK, CreateAPITokenResponse{
			ID:        record.ID,
			Name:      record.Name,
			TokenHint: record.TokenHint,
			Scopes:    record.Scopes,
			ExpiresAt: record.ExpiresAt,
			CreatedAt: record.CreatedAt,
			RawToken:  raw,
		})
	}
}

// writeJSON / writeJSONErr are shared by api_tokens.go + oauth.go in
// this package. Local copies (rather than reaching into internal/handlers)
// avoid an import cycle.
func writeJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(body)
}

func writeJSONErr(w http.ResponseWriter, status int, msg string) {
	typeName := "InternalServerException"
	switch status {
	case http.StatusBadRequest:
		typeName = "BadRequestException"
	case http.StatusUnauthorized:
		typeName = "AuthenticationException"
	case http.StatusForbidden:
		typeName = "ForbiddenException"
	case http.StatusNotFound:
		typeName = "NotFoundException"
	}
	writeJSON(w, status, map[string]any{
		"error": map[string]any{
			"message": msg,
			"type":    typeName,
			"code":    status,
		},
	})
}
